flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRowProcessFunction.scala

```diff
@@ -0,0 +1,157 @@
+import org.apache.flink.api.common.state.MapState
+import org.apache.flink.api.common.state.MapStateDescriptor
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.typeutils.ListTypeInfo
+import java.util.{ ArrayList, LinkedList, List => JList }
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo
+
+class BoundedProcessingOverRowProcessFunction(
+  private val aggregates: Array[AggregateFunction[_]],
+  private val aggFields: Array[Int],
+  private val bufferSize: Int,
+  private val forwardedFieldCount: Int,
+  private val aggregatesTypeInfo: RowTypeInfo,
+  private val inputType: TypeInformation[Row])
+    extends ProcessFunction[Row, Row] {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+  Preconditions.checkArgument(bufferSize > 0)
+
+  private var accumulators: Row = _
+  private var output: Row = _
+  private var accumulatorState: ValueState[Row] = _
+  private var rowMapState: MapState[Long, JList[Row]] = _
+  
+  override def open(config: Configuration) {
+    
+    output = new Row(forwardedFieldCount + aggregates.length)
+   // We keep the elements received in a list state 
+    // together with the ingestion time in the operator
+    val rowListTypeInfo: TypeInformation[JList[Row]] =
+      new ListTypeInfo[Row](inputType).asInstanceOf[TypeInformation[JList[Row]]]
+    
+    val mapStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]]("windowBufferMapState",
+        BasicTypeInfo.LONG_TYPE_INFO.asInstanceOf[TypeInformation[Long]], rowListTypeInfo)
+    
+    rowMapState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+    val stateDescriptor: ValueStateDescriptor[Row] =
+      new ValueStateDescriptor[Row]("aggregationState", aggregatesTypeInfo)
+      
+    accumulatorState = getRuntimeContext.getState(stateDescriptor)
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+    
+    val currentTime = ctx.timerService().currentProcessingTime()
+    var i = 0
+
+    var accumulators = accumulatorState.value()
+    // initialize state for the first processed element
+    if(accumulators == null){
+      accumulators = new Row(aggregates.length)
+      while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      }
+    }
+    
+    val keyIter = rowMapState.keys.iterator
+    var oldestTimeStamp = currentTime
+    var toRetract: JList[Row] = null
+    var currentKeyTime: Long = 0L
+    i = 0
+    while(keyIter.hasNext){
+      currentKeyTime = keyIter.next
+      i += rowMapState.get(currentKeyTime).size()
+      if(currentKeyTime <= oldestTimeStamp){
+        oldestTimeStamp = currentKeyTime
+        toRetract = rowMapState.get(currentKeyTime)
```

Comment:
We only need move `toRetract = rowMapState.get(currentKeyTime)` to loop out, then `toRetract = rowMapState.get(oldestTimeStamp)`.